File: deploy_webchat.py
import paramiko, os, sys, io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')

HOST = "109.69.22.112"
USER = "root"
PASS = "And716443@"
LOCAL_DIR = r"C:\Users\user\Desktop\cline\impulse-bot\webchat"
REMOTE_DIR = "/opt/impulse-bot/webchat"

def upload_dir(sftp, local, remote):
    try:
        sftp.stat(remote)
    except FileNotFoundError:
        sftp.mkdir(remote)
    for item in os.listdir(local):
        lp = os.path.join(local, item)
        rp = remote + "/" + item
        if os.path.isfile(lp):
            print(f"  Upload: {item}")
            sftp.put(lp, rp)
        elif os.path.isdir(lp):
            upload_dir(sftp, lp, rp)

def run_cmd(ssh, cmd):
    print(f"> {cmd}")
    stdin, stdout, stderr = ssh.exec_command(cmd, timeout=60)
    out = stdout.read().decode()
    err = stderr.read().decode()
    if out:
        print(out)
    if err:
        print(err)
    return out, err

print("Connecting to server...")
ssh = paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
ssh.connect(HOST, username=USER, password=PASS, timeout=15)
print("Connected!")

sftp = ssh.open_sftp()
print("Uploading webchat...")
upload_dir(sftp, LOCAL_DIR, REMOTE_DIR)
sftp.close()
print("Upload complete!")

run_cmd(ssh, f"cd {REMOTE_DIR} && npm install --production")
run_cmd(ssh, "pm2 delete impulse-webchat 2>/dev/null; cd /opt/impulse-bot && pm2 start webchat/server.js --name impulse-webchat")
run_cmd(ssh, "pm2 save")
out, _ = run_cmd(ssh, "curl -s http://localhost:3002/health")
print(f"\nHealth check: {out}")

run_cmd(ssh, "pm2 status")
ssh.close()
print("\nDone!")

File: deploy-webchat.js
const { Client } = require('ssh2');
const fs = require('fs');
const path = require('path');

const conn = new Client();
const serverContent = fs.readFileSync(path.join(__dirname, 'webchat', 'server.js'), 'utf-8');

conn.on('ready', () => {
  console.log('SSH connected!');
  
  // Upload the fixed file using SFTP
  conn.sftp((err, sftp) => {
    if (err) { console.error('SFTP error:', err); conn.end(); process.exit(1); return; }
    const ws = sftp.createWriteStream('/opt/impulse-bot/webchat/server.js', { mode: 0o644 });
    ws.on('close', () => {
      console.log('File uploaded via SFTP');
      
      sftp.end();
      // Restart PM2
      conn.exec('cd /opt/impulse-bot && pm2 restart impulse-webchat && sleep 2 && pm2 logs impulse-webchat --lines 10 --nostream', (e2, stream2) => {
        let out2 = '';
        stream2.on('data', d => out2 += d);
        stream2.stderr.on('data', d => out2 += d);
        stream2.on('close', () => {
          console.log('\n=== PM2 RESTART ===');
          console.log(out2);
          
          // Quick health test
          conn.exec('curl -s http://localhost:3002/health && echo "" && curl -s -X POST http://localhost:3002/socket.io/?EIO=4\\&transport=polling | head -c 200', (e3, stream3) => {
            let out3 = '';
            stream3.on('data', d => out3 += d);
            stream3.stderr.on('data', d => out3 += d);
            stream3.on('close', () => {
              console.log('\n=== HEALTH TEST ===');
              console.log(out3);
              conn.end();
              process.exit(0);
            });
          });
        });
      });
    });
    ws.write(serverContent);
    ws.end();
  });
});

conn.on('error', (err) => {
  console.error('SSH error:', err.message);
  process.exit(1);
});

console.log('Deploying fixed webchat server.js...');
conn.connect({
  host: '109.69.22.112',
  port: 22,
  username: 'root',
  password: 'And716443@',
  readyTimeout: 30000,
});